src/core/tsi/transport_security.cc

```diff
@@ -199,12 +204,22 @@ tsi_result tsi_handshaker_next(
     tsi_handshaker_on_next_done_cb cb, void* user_data) {
   if (self == nullptr || self->vtable == nullptr) return TSI_INVALID_ARGUMENT;
   if (self->handshaker_result_created) return TSI_FAILED_PRECONDITION;
+  if (self->handshake_shutdown) return TSI_HANDSHAKE_SHUTDOWN;
   if (self->vtable->next == nullptr) return TSI_UNIMPLEMENTED;
   return self->vtable->next(self, received_bytes, received_bytes_size,
                             bytes_to_send, bytes_to_send_size,
                             handshaker_result, cb, user_data);
 }
 
+void tsi_handshaker_shutdown(tsi_handshaker* self) {
+  if (self == nullptr || self->vtable == nullptr ||
```

Comment:
A minor improvement. In this way, for sync TSI, you don't have to define no-op shutdown. You can just set shutdown function as nullptr in vtable.

```
  if (self == nullptr || self->vtable == nullptr) {
    return;
  }
  self->handshake_shutdown = true;
  if (self->vtable->shutdown != nullptr) self->vtable->shutdown(self);
}
```

Reply:
Done.